pkg/templating/template.go

```diff
@@ -0,0 +1,114 @@
+package templating
+
+import (
+	"fmt"
+	"html/template"
+	"strings"
+
+	"github.com/Masterminds/sprig/v3"
+)
+
+// The split functions cannot be replaced with a single generic one, because there is
+// no common type between maps and slices
+
+// Given a map `src`, return two maps where the first contains key-value pairs that are not
+// templates, and the latter containers only key-value pairs that are templates.
+func SplitTemplatedMapItems(src map[string]string) (map[string]string, map[string]string) {
+	nonTemplateKeyPairs := make(map[string]string, len(src))
+	templateKeyPairs := make(map[string]string, len(src))
+
+	for key, value := range src {
+		if IsStringATemplate(value) {
+			templateKeyPairs[key] = value
+		} else {
+			nonTemplateKeyPairs[key] = value
+		}
+	}
+
+	return nonTemplateKeyPairs, templateKeyPairs
+}
+
+// Given a slice `src`, return two slices where the first contains key-value pairs that are not
+// templates, and the latter containers only key-value pairs that are templates. Values of the
+// opposing types will be set to an empty string, to preserve the position of all items.
+func SplitTemplatedListItems(src []string) ([]string, []string) {
+	nonTemplateKeyPairs := make([]string, len(src))
+	templateKeyPairs := make([]string, len(src))
+
+	for key, value := range src {
+		if IsStringATemplate(value) {
+			templateKeyPairs[key] = value
+		} else {
+			nonTemplateKeyPairs[key] = value
+		}
+	}
+
+	return nonTemplateKeyPairs, templateKeyPairs
+}
+
+// Returns true if the provided string contains a template, false otherwise
+func IsStringATemplate(str string) bool {
```

Comment:
This is more philosophical than technical, but should the following strings be considered malformed templates, or non-templates?
* `" .Node.Name }}"`
* `" {{ .Some.Value "`
* `"}} {{ .AnotherValue }}"`

I am of the opinion that if a string contains `{{` or `}}` (I missed this one), then it's probably better to assume that the user meant to use a template. There aren't a lot of cases where these values would be desired/used outside of templating.

What do you think? I'll defer to your opinion.

Reply:
See [here](https://github.com/budimanjojo/talhelper/pull/701/files#diff-9a39b1c30a3713cf232bac75ceb20106f7a47873969e9c2440cc246393045343R113-R168) for some test cases I wrote covering this

Reply:
I think using contains `{{` or `}}` will returns true for `only-opening`, `only-closing`, and `out-of-order` in your test? Because we should check if the index of first `{{` is there before the index of first `}}` and a string that only contains opening or closing is not a template?

Reply:
But you're right in a sense where it should error out in the templating function instead of being processed as a non-template and pass the labels/annotations validation.

Reply:
So just to make sure we're on the same page, any string containing one or more `{{` or one or more `}}` should be considered a template string, right? And if they `{{`/`}}` are in the wrong place, aren't paired, or are otherwise invalid, then they should still be considered a template string, which the templating library will reject, right?

Reply:
Yeah it makes sense. The downside is if the user really want to put something like only-opening, only-closing, or out-of-order scenario as value, I don't know if there's any use case that requires that but we can improve that in the future when that scenario happens IRL. Maybe you should also add a code comment about this quirk in case the future people looking at this code got confused.